File: practitioner-code/payments-python/src/payments/domain/currency.py
"""
Currency as an Enum so the type system itself forbids "USDD" or "us-dollar".
Each currency knows its own fraction digits - this drives Money's canonical
scale so equality is well-defined (see money.py).

Why an Enum (and not a str like "USD"):
  - IDE auto-completion + mypy can check exhaustiveness.
  - No "stringly typed" bugs ("usd" vs "USD" vs " USD ").
  - One place to add a new currency.
"""
from __future__ import annotations

from enum import Enum


class Currency(Enum):
    USD = ("USD", 2)
    EUR = ("EUR", 2)
    JPY = ("JPY", 0)   # yen has no minor unit - real bug source if hardcoded "2"

    def __init__(self, code: str, fraction_digits: int) -> None:
        self.code = code
        self.fraction_digits = fraction_digits

File: practitioner-code/payments-java/src/main/java/com/example/payments/domain/Currency.java
package com.example.payments.domain;

/**
 * Currency as an enum so the type system itself forbids "USDD" or "us-dollar".
 * Each currency knows its own fraction digits - this drives Money's canonical
 * scale so equality is well-defined (see Money).
 *
 * Why an enum (and not a String like "USD"):
 *  - Compile-time exhaustiveness in switch.
 *  - No "stringly typed" bugs ("usd" vs "USD" vs " USD ").
 *  - One place to add a new currency.
 */
public enum Currency {
    USD(2),
    EUR(2),
    JPY(0);   // yen has no minor unit - a real source of bugs if you hardcode "2"

    private final int fractionDigits;

    Currency(int fractionDigits) {
        this.fractionDigits = fractionDigits;
    }

    public int fractionDigits() {
        return fractionDigits;
    }
}
